src/Storages/VirtualColumnUtils.cpp

```diff
@@ -115,7 +119,29 @@ NameSet getVirtualNamesForFileLikeStorage()
     return {"_path", "_file", "_size", "_time"};
 }
 
-VirtualColumnsDescription getVirtualsForFileLikeStorage(const ColumnsDescription & storage_columns)
+std::unordered_map<std::string, std::string> parseHivePartitioningKeysAndValues(const std::string& path, const ColumnsDescription & storage_columns)
+{
+    std::string pattern = "/([^/]+)=([^/]+)";
+    re2::StringPiece input_piece(path);
+
+    std::unordered_map<std::string, std::string> key_values;
+    std::string key, value;
+    std::set<String> used_keys;
```

Comment:
For such case use `std::unordered_set`, because it's a hash table, so all operations are O(1). `std::set` is a rb-tree and all operations are O(log N)